### PR TITLE
Fix: Support for user-defined module fmt for all compilers

### LINKED ISSUE #5283: Reserved keywords in Go in need of escape
### Dafny version

latest-nightly

### Code to produce this issue

```dafny
module fmt {}

method Main(){
    print "done\n";
}
```

### Command to run and resulting output

Try to compile and run

### What happened?

It fails.
Follow-up from https://github.com/dafny-lang/dafny/issues/4181

### What type of operating system are you experiencing the problem on?

Windows

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR fixes #5283
I added the corresponding test.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Comment by fabiomadge** [comment]
Yes, this fixed the issue you opened, but the underlying problem still persists. Here's another problematic program.
```Dafny
module time {}

method Main(){
     print "done\n";
}
```
They are all reserved by the [GO standard library](https://pkg.go.dev/std). I would prefer to merge a more comprehensive fix.

File: docs/dev/news/5283.fix [hunk 1]
```diff
@@ -0,0 +1 @@
+Support for user-defined module fmt for all compilers
```
Comment (MikaelMayer) [inline docs/dev/news/5283.fix:1]:
```suggestion
Support for user-defined module with the same name as Go built-in modules, checked for all compilers
```

File: Source/DafnyCore/Backends/GoLang/GoCodeGenerator.cs [hunk 2]
```diff
@@ -2573,11 +2573,77 @@ protected class ClassWriter : IClassWriter {
...
+            res = ",\n      " + res;
+        } else if(i > 0) {
+            res = "," + res;
+            lineLength += res.length;
+        }
+        return res;
+    }).join(""))
+     */
+    public readonly HashSet<string> ReservedModuleNames = new() {
+      "C",
```
Comment (MikaelMayer) [inline Source/DafnyCore/Backends/GoLang/GoCodeGenerator.cs:2593]:
```suggestion
      "c",
```

File: Source/DafnyCore/Backends/Python/PythonCodeGenerator.cs [hunk 3]
```diff
@@ -686,8 +686,11 @@ private class ClassWriter : IClassWriter {
     }
 
     private string FullName(TopLevelDecl decl) {
-      var localDefinition = decl.EnclosingModuleDefinition == enclosingModule;
-      return IdProtect(localDefinition ? decl.GetCompileName(Options) : decl.GetFullCompileName(Options));
```
[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Backends/Python/PythonCodeGenerator.cs:690]: I don't know how to evaluate if this is good or not. I'll let the CI test this new behavior.

**Review by fabiomadge — DISMISSED** [review]
Thanks!

**Review by fabiomadge — APPROVED** [review]

**Pull request merged**